Feature/parquet arrow renderer

kyrre opened the pull request:
This adds Parquet and arrow renderers and closes #1860

kyrre committed:
add arrow and parquet renderers

kyrre committed:
formatting with ruff

kyrre committed:
linting

machinelearningdesignpatterns committed:
Fix: Avoid import-time errors if pyarrow is missing

- Use string type annotations for pa.* types in Arrow/Parquet renderers.
- Ensure pyarrow dependency is checked only in __init__.

machinelearningdesignpatterns committed:
fix: handle layerdata to pyarrow conversion

ikelos reviewed (commented):
Thanks very much, this looks really nice.  I still want to check through a few more bits and pieces, but generally it looks good.

At the moment `text_renderer` is looking a little full though, I was wondering if you could test whether this still works, if all the new classes are moved into a file like `volatility3/plugins/parquet_renderer.py` or similar?  We should automatically import all python files under plugins, and I believe we do it before we choose any renderers, which means it should still show up in the available options.  We'll probably want to make a renderers directory that people can drop things into, this would just allow users to drop in their own renderers (and might therefore act as a good example of how to write new ones when people want).

Anyway, just some thoughts.  I'll give it a full review over the coming week and add in any comments of anything else, but I'm really pleased someone else has written a renderer for a very different format finally!  Thanks!  5:)

kyrre commented:
> At the moment `text_renderer` is looking a little full though, I was wondering if you could test whether this still works, if all the new classes are moved into a file like `volatility3/plugins/parquet_renderer.py` or similar? We should automatically import all python files under plugins, and I believe we do it before we choose any renderers, which means it should still show up in the available options. We'll probably want to make a renderers directory that people can drop things into, this would just allow users to drop in their own renderers (and might therefore act as a good example of how to write new ones when people want).

I really like the structure you're proposing here. However, it does not work at the moment because the plugins are imported after the list of available renderers is generated:

https://github.com/volatilityfoundation/volatility3/blob/997f2a4f4c32cbb268ae7236ab3a7e24a3ec9bb2/volatility3/cli/__init__.py#L109-L115

https://github.com/volatilityfoundation/volatility3/blob/997f2a4f4c32cbb268ae7236ab3a7e24a3ec9bb2/volatility3/cli/__init__.py#L196-L203

and then

https://github.com/volatilityfoundation/volatility3/blob/997f2a4f4c32cbb268ae7236ab3a7e24a3ec9bb2/volatility3/cli/__init__.py#L336-L338

ikelos commented:
Ok, thanks for investigating, I'll take a look into it...  5:)

On Mon, 30 Jun 2025, 10:34 Kyrre Wahl Kongsgård, ***@***.***>
wrote:

> *kyrre* left a comment (volatilityfoundation/volatility3#1861)
> <https://github.com/volatilityfoundation/volatility3/pull/1861#issuecomment-3018461421>
>
> At the moment text_renderer is looking a little full though, I was
> wondering if you could test whether this still works, if all the new
> classes are moved into a file like volatility3/plugins/parquet_renderer.py
> or similar? We should automatically import all python files under plugins,
> and I believe we do it before we choose any renderers, which means it
> should still show up in the available options. We'll probably want to make
> a renderers directory that people can drop things into, this would just
> allow users to drop in their own renderers (and might therefore act as a
> good example of how to write new ones when people want).
>
> I really like the structure you're proposing here. However, it does not
> work at the moment because the plugins are imported after the list of
> available renderers is generated:
>
>
> https://github.com/volatilityfoundation/volatility3/blob/997f2a4f4c32cbb268ae7236ab3a7e24a3ec9bb2/volatility3/cli/__init__.py#L109-L115
>
>
> https://github.com/volatilityfoundation/volatility3/blob/997f2a4f4c32cbb268ae7236ab3a7e24a3ec9bb2/volatility3/cli/__init__.py#L196-L203
>
> and then
>
>
> https://github.com/volatilityfoundation/volatility3/blob/997f2a4f4c32cbb268ae7236ab3a7e24a3ec9bb2/volatility3/cli/__init__.py#L336-L338
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/volatilityfoundation/volatility3/pull/1861#issuecomment-3018461421>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALIZVNJJF4OCEYJABADI2L3GEACFAVCNFSM6AAAAACAJHP2PKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAMJYGQ3DCNBSGE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

machinelearningdesignpatterns committed:
moved arrow/parquet parsers to it's own file and reverted changes

machinelearningdesignpatterns committed:
moved renderer parsing to after the plugins are imported

machinelearningdesignpatterns committed:
black formatting

machinelearningdesignpatterns committed:
restore orginal text_renderer

kyrre commented:
> Ok, thanks for investigating, I'll take a look into it... 5:)

I created a new file in the plugin directory `parquet_renderer.py` and moved all the code related to the renderers in `cli/__init__.py` to after the plugins have been imported. What I moved:

- creating the list of renderers
- cli parsing and validation for the renderer arg
- display banner output redirection 

How do you feel about this @ikelos - is this something you can work with? I agree that it is cleaner to create a renderer directory, but maybe this is OK for now? 🤔

machinelearningdesignpatterns committed:
fix bug: hadn't properly renamed the write_data method name

ikelos commented on volatility3/framework/plugins/parquet_renderer.py:
```diff
@@ -0,0 +1,166 @@
+vollog = logging.getLogger(__name__)
+
+try:
+    ARROW_PRESENT = True
```
I'd probably set ARROW_PRESENT to False outside/before the try, and then only when everything's imported successfully set it to true.  This way round should work, but feels like there could more easily be a failure of some kind.  It's not a show stopper though.  We *probably* should make sure this stuff is versioned though, in case someone decides they want to extend it, or a future version changes the output in some way...  5:S

kyrre replied:
> I'd probably set ARROW_PRESENT to False outside/before the try, and then only when everything's imported successfully set it to true.

Yes, that definitely feels more safe. I have updated the file.

When it comes to versioning - what did you have in mind? Other files seem to use a tuple class attribute, 

`_version = (1, 0, 0)` 

I added this attribute to both he arrow and parquet renderer classes.

ikelos reviewed (commented):
Sorry to be so indecisive (and to take so long to rereview it).  I think it's much closer and I'm pretty happy with the changes to the main CLI, which is likely the biggest hurdle.  The trick now is to get this renderer right (since it'll likely be used as the template for everybody's future renderers).  As such, could we try putting it inside of volatility3/plugins/renderers please?  I don't *think* there's any functionality that will decide it's only for `renderer` OSes, but best to check it now rather than trying to change it later.  Other than that and the other comment I posted, I think it could be ready for prime time?

machinelearningdesignpatterns committed:
moved the arrow/parquet renderers to a subdir for renderers

machinelearningdesignpatterns committed:
remove old parquet renderer

machinelearningdesignpatterns committed:
fix: update data type mappings for format_hints.Bin and format_hints.Hex to pa.uint64

kyrre commented on volatility3/framework/plugins/renderers/parquet_renderer.py:
I moved the parquet and arrow renderers into this file as requested, and there were no issues with the cli discovering them.

kyrre commented on volatility3/framework/plugins/renderers/parquet_renderer.py:
```diff
@@ -0,0 +1,173 @@
+            float: pa.float64,
+            str: pa.utf8,
+            datetime.datetime: lambda: pa.timestamp("ms"),
+            format_hints.Bin: pa.uint64,
```
When running all the windows plugins locally, I noticed that there were some integer overflow issues, and so I changed the `format_hints.Bin` and `format_hints.Hex` to map to `pa.uint64` instead.

kyrre commented:
> Sorry to be so indecisive (and to take so long to rereview it). I think it's much closer and I'm pretty happy with the changes to the main CLI, which is likely the biggest hurdle. The trick now is to get this renderer right (since it'll likely be used as the template for everybody's future renderers). As such, could we try putting it inside of volatility3/plugins/renderers please? I don't _think_ there's any functionality that will decide it's only for `renderer` OSes, but best to check it now rather than trying to change it later. Other than that and the other comment I posted, I think it could be ready for prime time?

I think we should be set! What do you think @ikelos ? Another thing we could add is a test suite perhaps 🤔

machinelearningdesignpatterns committed:
Added tests for the Parquet and Arrow renderers

kyrre commented:
I went ahead and added some tests! 

The testing strategy I choose was to port a few of the plugin tests and express the assertions using a [PyArrow table](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html) that was instantiated from both Parquet/Arrow data emitted to stdout by the plugins.

machinelearningdesignpatterns committed:
flatten tree output

kyrre commented:
It didn't work correctly with plugins that created a nested output, e.g., windows.pstree. I fixed this by flattening and adding _vol_id and _vol_parent_id columns to the arrow schema.

machinelearningdesignpatterns committed:
black & ruff

dgmcdona commented on volatility3/framework/plugins/renderers/parquet_renderer.py:
```diff
@@ -0,0 +1,221 @@
+
+        return pa.schema(fields)
+
+    def _flatten_tree_structure(self, nested: list[dict]) -> list[dict]:
```
The minimum Python version for volatility3 is 3.8, so these hints need to use `typing.List` and `typing.Dict`.

kyrre replied:
Thank you for the review! I have updated the type hints now.

dgmcdona commented on volatility3/framework/plugins/renderers/parquet_renderer.py:
```diff
@@ -0,0 +1,221 @@
+    _version = (1, 0, 0)
+
+    def __init__(
+        self, options: List[interfaces.renderers.RenderOption] | None = None
```
For Python 3.8 support, we'll need this to be `typing.Optional` instead of using the pipe syntax.

dgmcdona commented on volatility3/framework/plugins/renderers/parquet_renderer.py:
```diff
@@ -0,0 +1,221 @@
+        rows = []
+        self._node_id_counter = 0
+
+        def _process_node(node: dict, parent_id: int | None):
```
Convert to `typing.Optional`

machinelearningdesignpatterns committed:
fix: update type hints to support Python 3.8

ikelos commented on volatility3/framework/plugins/renderers/parquet_renderer.py:
```diff
@@ -0,0 +1,222 @@
+    CLIRenderer,
+    display_disassembly,
+    LayerDataRenderer,
+)
```
```suggestion
from volatility3.cli import text_renderer
```

The [volatility coding style](https://github.com/volatilityfoundation/volatility3/blob/develop/CODING_STYLE.md#imports) is to import only modules, so that people don't import a class directly from a module where it had been imported itself.  Please could you update these.  We allow it for typing because they're often repeated and very common.

ikelos commented on volatility3/framework/plugins/renderers/parquet_renderer.py:
```diff
@@ -0,0 +1,222 @@
+    vollog.debug("Arrow/Parquet libraries not found")
+
+
+class ArrowRenderer(CLIRenderer):
```
```suggestion
class ArrowRenderer(text_renderer.CLIRenderer):
```

ikelos commented on volatility3/framework/plugins/renderers/parquet_renderer.py:
```diff
@@ -0,0 +1,222 @@
+                    data = None
+
+                if isinstance(data, renderers.Disassembly):
+                    data = display_disassembly(data)
```
```suggestion
                    data = text_renderer.display_disassembly(data)
```

ikelos commented on volatility3/framework/plugins/renderers/parquet_renderer.py:
```diff
@@ -0,0 +1,222 @@
+                    data = display_disassembly(data)
+
+                if isinstance(data, renderers.LayerData):
+                    data = LayerDataRenderer().render_bytes(data)[0]
```
```suggestion
                    data = text_renderer.LayerDataRenderer().render_bytes(data)[0]
```

ikelos reviewed (commented):
Thanks for holding with us through the delay.  This looks good now, just a couple more comments and then it should be able to land.  5:)

kyrre committed:
Update volatility3/framework/plugins/renderers/parquet_renderer.py

Co-authored-by: ikelos <ikelos@users.noreply.github.com>

kyrre committed:
Update volatility3/framework/plugins/renderers/parquet_renderer.py

Co-authored-by: ikelos <ikelos@users.noreply.github.com>

kyrre committed:
Update volatility3/framework/plugins/renderers/parquet_renderer.py

Co-authored-by: ikelos <ikelos@users.noreply.github.com>

kyrre committed:
Update volatility3/framework/plugins/renderers/parquet_renderer.py

Co-authored-by: ikelos <ikelos@users.noreply.github.com>

ikelos commented on test/renderers/test_parquet_renderers.py:
```diff
@@ -0,0 +1,108 @@
+    import pyarrow.parquet as pq
+    import pyarrow.compute as pc
+    HAS_PYARROW = True
+except ImportError:
```
```suggestion
except ImportError:
   # The user doesn't have pyarrow installed, but HAS_PYARROW will be false so just continue
```
Last change to get it over the line...  5:)

kyrre replied:
Thank you 😁

ikelos committed:
Update test/renderers/test_parquet_renderers.py
